### ISSUE TITLE
[Idea] "broker-lite" to store only the missing files

### ISSUE THREAD

**Issue by digizeph** [bgpkit/bgpkit-broker#39]
## TL;DR

A could-friendly version of BGPKIT Broker that **_stores only the missing files_** and use cloud platforms to store and serve the data queries.

## Background and Idea

The files in RIPE RIS and RouteViews is stored in a very consistent way since the very beginning of these projects:
* RIS: 8-hour RIB dump, 5-minute updates
* RouteViews: 2-hour RIB dump, 15-minute updates

Currently, we crawl and fetch all files' meta data and store the information in a database, which ended up being around 40+ million entries in size. Although we managed to store that in a SQLite database successfully, we could potentially shrink down the information further. The even smaller database can fit in more storage locations and have even greater potential for adoption. For example, 40 million entries would unlikely to fit in a single edge database like D1.

Instead of storing "what's in the archives," we could instead storing "what's missing in the archives." For example, for a time range, we have expectation on the files "should" be available in the range. If outages happen and archives missing certain files, we should store the missing files information in the DB. During lookup time, we would need to go fetch the start and latest timestamp for each collector, making sure they have available data for the range, and then reconstruct available files based on missing files information.

## Data Storage

We would have the following tables.

### `collectors`: route collector's basic information

- `name`: collector name
- `project`: `routeviews` or `riperis`
- `rib_base_url`: base URL for RIB dump files
- `updates_base_url`: base URL for updates files
- `rib_interval`: time interval expected between each RIB dump file in seconds
- `updates_interval`: time interval expected between each updates file in seconds

### `earliest_latest`: earliest and latest file timestamps

- `name`: collector name
- `earliest_rib_ts`: earliest RIB file timestamp
- `earliest_updates_ts`: earliest updates file timestamp
- `latest_rib_ts`: latest RIB file timestamp
- `latest_updates_ts`: latest updates file timestamp

### `missing_files`: missing files

- `collector`: collector name
- `data_type`: RIB or updates
- `timestamp`: timestamp of the missing file

### Storage and API platform

This lighter version of files are great for store on cloud edge platforms like Cloudflare D1 or Turso. We can also implement WASM/JS APIs directly on the cloud platforms instead of having to deploy on dedicated baremetal or VPS hardware.

### Crawler

We have previously experimented with the idea of running a crawler directly from Cloudflare Workers. Although it was generally OK, I find it still quite slow and prone to unhandled errors. Maybe we can revisit this idea now. Having both the crawler and storage on a cloud platform will free us from the deployment concerns, but would make the project like to be cloud-platform dependent, which is not ideal. We will have to continue investigate this topic.

### Bootstrapping

We can utilize the current 0.7 beta SQLite database to bootstrap this new DB. Should be straightforward to achieve. Perhaps, for the long run, we could have the "full" version and the "lite" version (or "cloud") version running in the same time.
